### ISSUE TITLE
Broken on recent nightlies

### ISSUE THREAD

**Issue by bsteinb** [bsteinb/rsmpi#45]
`mpi` does not build on recent nightlies, I have not yet tried to figured out when this first broke. This is what is broken:

```
error[E0308]: mismatched types
  --> src/request.rs:65:1
   |
65 | / pub struct Request<'a, S: Scope<'a> = StaticScope> {
66 | |     request: MPI_Request,
67 | |     scope: S,
68 | |     phantom: PhantomData<Cell<&'a ()>>,
69 | | }
   | |_^ lifetime mismatch
   |
   = note: expected type `request::Scope<'a>`
              found type `request::Scope<'static>`
note: the lifetime 'a as defined on the struct at 65:1...
  --> src/request.rs:65:1
   |
65 | pub struct Request<'a, S: Scope<'a> = StaticScope> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
   --> src/request.rs:230:1
    |
230 | pub struct WaitGuard<'a, S: Scope<'a> = StaticScope>(Option<Request<'a, S>>);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ lifetime mismatch
    |
    = note: expected type `request::Scope<'a>`
               found type `request::Scope<'static>`
note: the lifetime 'a as defined on the struct at 230:1...
   --> src/request.rs:230:1
    |
230 | pub struct WaitGuard<'a, S: Scope<'a> = StaticScope>(Option<Request<'a, S>>);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: ...does not necessarily outlive the static lifetime

error[E0308]: mismatched types
   --> src/request.rs:274:1
    |
274 | pub struct CancelGuard<'a, S: Scope<'a> = StaticScope>(WaitGuard<'a, S>);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ lifetime mismatch
    |
    = note: expected type `request::Scope<'a>`
               found type `request::Scope<'static>`
note: the lifetime 'a as defined on the struct at 274:1...
   --> src/request.rs:274:1
    |
274 | pub struct CancelGuard<'a, S: Scope<'a> = StaticScope>(WaitGuard<'a, S>);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: ...does not necessarily outlive the static lifetime

error: aborting due to 3 previous errors
```

@Rufflewind any idea what this might be?

**Comment by Rufflewind** [comment]
It seems the compiler no longer allows default parameters to constrain other parameters?

I don't know what change caused this (NLL?), but given this had been working previously it would be useful to file an issue with the compiler and see if this was an intentional breaking change.

If this is intentional, we could try working around by removing the default parameter and just always be explicit.

**Comment by bsteinb** [comment]
Okay, I will try to bisect the nightly that broke this.

**Comment by bsteinb** [comment]
Opened an issue: https://github.com/rust-lang/rust/issues/49344

**Comment by bsteinb** [comment]
Just released `mpi-0.5.4` which works around this issue.

**Comment by bsteinb** [comment]
Yeah, I know, I followed the issue over at rust-lang/rust. I left this issue open until I could find the time to revert the workaround.
